plugins/logitech-scribe/fu-logitech-scribe-device.c

```diff
@@ -0,0 +1,603 @@
+	fu_byte_array_append_uint32(end_pkt, 0, G_LITTLE_ENDIAN);   /* force */
+	fu_byte_array_append_uint32(end_pkt, MD5, G_LITTLE_ENDIAN); /* checksum type */
+	g_byte_array_append(end_pkt, (const guint8 *)base64hash, strlen(base64hash));
+	if (!fu_logitech_scribe_device_send_upd_cmd(self, CMD_END_TRANSFER, end_pkt, error)) {
+		g_prefix_error(error, "failed to write end transfer transfer packet: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* uninitialize */
+	/* no need to wait for ACK message, perhaps device reboot already in progress, ignore */
+	if (!fu_logitech_scribe_device_send_upd_cmd(self, CMD_UNINIT, NULL, &error_local)) {
+		g_debug("Failed to receive acknowledgment for uninitialize request, ignoring it");
+	}
+	fu_progress_step_done(progress);
+
+	/*
+	 * image file pushed. Device validates and uploads new image on inactive partition. Reboots
+	 * wait for RemoveDelay duration
+	 */
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_scribe_device_set_version(FuDevice *device, GError **error)
+{
+	FuLogitechScribeDevice *self = FU_LOGITECH_SCRIBE_DEVICE(device);
+	g_autofree gchar *fwversion_str = NULL;
+	guint32 fwversion = 0;
+	guint16 data_len = 0;
+	g_autofree guint8 *query_data = NULL;
+
+	/* query current device version */
+	if (!fu_logitech_scribe_device_query_data_size(self,
+						       kLogiUnitIdCameraVersion,
+						       kLogiCameraVersionSelector,
+						       &data_len,
+						       error))
+		return FALSE;
+	query_data = g_malloc0(data_len);
+	if (!fu_logitech_scribe_device_get_xu_control(self,
+						      kLogiUnitIdCameraVersion,
+						      kLogiCameraVersionSelector,
+						      data_len,
+						      (guchar *)query_data,
+						      error))
+		return FALSE;
+
+	/*  little-endian data. MinorVersion byte 0, MajorVersion byte 1, BuildVersion byte 3 & 2 */
+	fwversion =
+	    (query_data[1] << 24) + (query_data[0] << 16) + (query_data[3] << 8) + query_data[2];
+	fwversion_str = fu_version_from_uint32(fwversion, FWUPD_VERSION_FORMAT_TRIPLET);
+	fu_device_set_version(device, fwversion_str);
+	g_debug("Version data: %u, Version string: %s ", fwversion, fwversion_str);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_scribe_device_setup(FuDevice *device, GError **error)
+{
+	FuLogitechScribeDevice *self = FU_LOGITECH_SCRIBE_DEVICE(device);
+	g_autoptr(GPtrArray) intfs = NULL;
+	g_autoptr(GUsbDevice) g_usb_device = NULL;
+
+	if (g_getenv("FWUPD_LOGITECH_SCRIBE_VERBOSE") != NULL) {
+		g_debug("Processing setup");
+	}
+
+	if (!fu_logitech_scribe_device_set_version(device, error))
+		return FALSE;
+
+	/* convert GUdevDevice to GUsbDevice */
+	g_usb_device = fu_udev_device_find_usb_device(FU_UDEV_DEVICE(device), error);
+	if (g_usb_device == NULL)
+		return FALSE;
+	self->usb_device = fu_usb_device_new(fu_device_get_context(device), g_usb_device);
+	if (self->usb_device == NULL)
+		return FALSE;
```

Comment:
I don't think we should store the `usb_device` in the class private -- potentially the GUsbDevice and GUdevDevice have different lifecycles (e.g. at unplug) and so I think we should get the GUsbDevice every time it's needed.